kolibri/plugins/coach/class_summary_api.py

```diff
@@ -19,23 +20,82 @@
 COMPLETED = "completed"
 
 
-class ContentStatusSerializer(serializers.ModelSerializer):
-    status = serializers.SerializerMethodField()
-    learner_id = serializers.PrimaryKeyRelatedField(source="user", read_only=True)
-    last_activity = serializers.CharField(source="end_timestamp")
-
-    def get_status(self, content_summary_log):
-        # if content_summary_log.help_needed:
-        #     return HELP_NEEDED
-        if content_summary_log.progress == 1:
+def content_status_serializer(lesson_data, learners_data, classroom):
+
+    # First generate a unique set of content node ids from all the lessons
+    lesson_node_ids = set()
+    for lesson in lesson_data:
+        lesson_node_ids |= set(lesson.get("node_ids"))
+
+    # Now create a map of content_id to node_id so that we can map between lessons, and notifications
+    # which use the node id, and summary logs, which use content_id
+    content_map = {n[0]: n[1] for n in ContentNode.objects.filter(id__in=lesson_node_ids).values_list("content_id", "id")}
+
+    # Get all the values we need from the summary logs to be able to summarize current status on the
+    # relevant content items.
+    content_log_values = logger_models.ContentSummaryLog.objects.filter(
+        content_id__in=set(content_map.keys()), user__in=[learner["id"] for learner in learners_data]
+    ).values("user_id", "content_id", "end_timestamp", "time_spent", "progress")
+
+    # In order to make the lookup speedy, generate a unique key for each user/node that we find
+    # listed in the needs help notifications that are relevant. We can then just check
+    # existence of this key in the set in order to see whether this user has been flagged as needing
+    # help.
+    lookup_key = "{user_id}-{node_id}"
+    needs_help = {
+        lookup_key.format(user_id=n[0], node_id=n[1]): n[2] for n in LearnerProgressNotification.objects.filter(
+            classroom_id=classroom.id,
+            notification_event=NotificationEventType.Help,
+            lesson_id__in=[lesson["id"] for lesson in lesson_data],
+        ).values_list("user_id", "contentnode_id", "timestamp")
+    }
```

Comment:
These dictionary comprehensions and tuple lookups obfuscate the intended logic. Some intermediary variables would help self-document the code:

```python
learner_tuples = LearnerProgressNotification.objects.filter(
    classroom_id=classroom.id,
    notification_event=NotificationEventType.Help,
    lesson_id__in=[lesson["id"] for lesson in lesson_data],
).values_list("user_id", "contentnode_id", "timestamp")

needs_help = {}
for (user_id, contentnode_id, timestamp) in learner_tuples:
    needs_help["{}-{}".format(user_id, contentnode_id)] = timestamp
```

---

Side note - I like the strategy of generating a new key. In vuex I used nested objects, but this seems cleaner